src/lib/storage/column_accessor.hpp

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+
+#include "resolve_type.hpp"
+#include "storage/reference_column.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+/**
+ * This is the base class for all ColumnAccessor types.
+ * It provides the common interface to access individual values of a column.
+ */
+template <typename T>
+class BaseColumnAccessor {
+ public:
+  virtual const std::optional<T> access(ChunkOffset offset) const = 0;
+  virtual ~BaseColumnAccessor() {}
+};
+
+/**
+ * A ColumnAccessor is templated per ColumnType and DataType (T).
+ * It requires that the underlying column implements an implicit interface:
+ *
+ *   const std::optional<T> get_typed_value(const ChunkOffset chunk_offset) const;
+ *
+ */
+template <typename T, typename ColumnType>
+class ColumnAccessor : public BaseColumnAccessor<T> {
+ public:
+  explicit ColumnAccessor(const ColumnType& column) : _column{column} {}
+
+  const std::optional<T> access(ChunkOffset offset) const final { return _column.get_typed_value(offset); }
+
+ protected:
+  const ColumnType& _column;
+};
+
+/**
+ * Partial template specialization for ReferenceColumns.
+ * Since ReferenceColumns don't know their 'T', this uses the subscript operator as a fallback.
+ */
+template <typename T>
+class ColumnAccessor<T, ReferenceColumn> : public BaseColumnAccessor<T> {
+ public:
+  explicit ColumnAccessor(const ReferenceColumn& column) : _column{column} {}
+  const std::optional<T> access(ChunkOffset offset) const final {
+    const auto all_type_variant = _column[offset];
```

Comment:
A ColumnAccessor should never be created on a ReferenceColumn, am I right? Just `Fail()` here, if I am.

Reply:
The motivating use case for ColumnAccessors is the ReferenceColumnIterable and hopefully that will not try to access other ReferenceColumns, correct. However, I don't see why this should fail if it works just fine and isn't slower than the usual `operator[]`.

Reply:
Maybe PerformanceWarning as middle ground?

Reply:
done